Browser extension provider communication (#1)
jiexi/CAIPs: caip-browser_extension_provider_communication.md

## Simple Summary
<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the CAIP.-->
This CAIP discusses the motivation, specification, and rationale for a proposal aimed at improving how web extension wallets interact with websites. It outlines the current method of injecting JavaScript provider APIs into websites, its advantages, and its numerous disadvantages, such as security concerns, performance issues, and the risk of breaking websites. An alternative strategy is proposed that specifies a standard communication specification over a new transport layer which enables websites to be able to embed their own provider as a library, addressing the disadvantages of injecting providers into websites and improving web extension interoperability as a whole.

## Abstract
<!--A short (~200 word) description of the technical issue being addressed.-->

> In the current web extension wallet ecosystem, the prevalent approach involves injecting a JavaScript provider API, such as `window.ethereum`, directly into websites as a global variable. This method offers simplicity for web developers and ensures compatibility across extensions but raises significant concerns regarding security, performance, and potential disruption of website functionality due to the need for extensive permissions and the injection of additional code. An alternative approach involves websites embedding their own provider libraries, which could mitigate these issues by reducing required permissions, enhancing performance, and providing developers with greater control over provider integration. To make this approach feasible, a communication protocol must be standardized for the purpose.

Reviewer: We say "the prevalent approach" here without saying what it's the approach _for_. Maybe obvious from context, but we can probably word this better. e.g. something like "In the current web extension wallet ecosystem, most wallets communicate with websites by injecting a JavaScript provider API (`window.ethereum`) directly into websites as a global variable.

> In the current web extension wallet ecosystem, the prevalent approach involves injecting a JavaScript provider API, such as `window.ethereum`, directly into websites as a global variable. This method offers simplicity for web developers and ensures compatibility across extensions but raises significant concerns regarding security, performance, and potential disruption of website functionality due to the need for extensive permissions and the injection of additional code. An alternative approach involves websites embedding their own provider libraries, which could mitigate these issues by reducing required permissions, enhancing performance, and providing developers with greater control over provider integration. To make this approach feasible, a communication protocol must be standardized for the purpose.

Reviewer: >potential disruption of website functionality due to the need for extensive permissions and the injection of additional code

Hmm. I understand what "potential disruption of website functionality" could mean, but I don't see a clear connection between that and extensive permissions. The website functionality disruptions I'm aware of are due to contentscript messaging and code injection.

> In the current web extension wallet ecosystem, the prevalent approach involves injecting a JavaScript provider API, such as `window.ethereum`, directly into websites as a global variable. This method offers simplicity for web developers and ensures compatibility across extensions but raises significant concerns regarding security, performance, and potential disruption of website functionality due to the need for extensive permissions and the injection of additional code. An alternative approach involves websites embedding their own provider libraries, which could mitigate these issues by reducing required permissions, enhancing performance, and providing developers with greater control over provider integration. To make this approach feasible, a communication protocol must be standardized for the purpose.

Reviewer: >To make this approach feasible, a communication protocol must be standardized for the purpose.

Hmm, something seems wrong grammatically here (for "the" purpose?). Maybe we're better off deleting this sentence, the next paragraph covers this more effectively.

> In the current web extension wallet ecosystem, the prevalent approach involves injecting a JavaScript provider API, such as `window.ethereum`, directly into websites as a global variable. This method offers simplicity for web developers and ensures compatibility across extensions but raises significant concerns regarding security, performance, and potential disruption of website functionality due to the need for extensive permissions and the injection of additional code. An alternative approach involves websites embedding their own provider libraries, which could mitigate these issues by reducing required permissions, enhancing performance, and providing developers with greater control over provider integration. To make this approach feasible, a communication protocol must be standardized for the purpose.

Reviewer 2: I think we should mention that CAIP-282 already addresses the issue with window.ethereum legacy and that this proposal is enabling that standard to be transported through externally_connectable

> This proposal addresses the challenge of maintaining interoperability and extensibility between web extensions and websites by standardizing the communication protocol through the `externally_connectable` interface. This standardization aims to facilitate seamless interaction across different web extensions, ensuring a consistent and secure method of communication. This proposal outlines a specific message format for this communication and discusses the use of `externally_connectable` to allow web extensions to send and receive messages with authorized websites and extensions.

Reviewer: A few notes:
* Do we need to say "and extensibility" here? I am not sure what that part means. If it's referring to how the protocol we're proposing here is itself extensible, maybe it would be clearer to say "This proposal addresses the challenge of maintaining interoperability between web extensions and websites by standardizing an extensible communication protocol through the `externally_connectable` interface
* Some of the language here seems a bit "fluffy" or like marketing speak. "facilitate seamless interaction", "ensuring a consistent and secure method of communication". I don't know what we're trying to communicate with those.
* "authorized websites and extensions"? This protocol has no authorization.

> Despite the lack of current support for `externally_connectable` in Firefox, the proposal underscores the importance of interoperability and standardized communication for the future of web extension wallets, advocating for a transition away from contentscript injection.

Reviewer: "underscores the importance of ____" and "advocating for ____" seems like meta-commentary. We don't need to advocate for things or comment on what our proposal underscores the importance of from within the proposal itself.

We're also undermining our own point here before we make it, by bringing up Firefox compatibility. That is definitely worth mentioning in the abstract, but we can be a bit less self-defeating in how we do that.

e.g. maybe "This proposal is not yet applicable to Firefox due to lack of support for `externally_connectable`, but they are considering implementing it". That's short, communicates the caveat affectively, but with an optimistic (and still accurate) framing.

## Specification
<!--The technical specification should describe the standard in detail. The specification should be detailed enough to allow competing, interoperable implementations. -->

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
uppercase in this document are to be interpreted as described in [RFC
2119](https://www.ietf.org/rfc/rfc2119.txt)

### Summary

Web extensions should expose a standard interface over [`externally_connectable`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable) to enable the inter-operability of embedded providers.

### Message format

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "/caip-x/schemas/provider-request.schema.json",
  "title": "Provider / Web Extension message",
  "description": "A request sent between a provider and a web extension.",
  "type": "object",
  "properties": {
    "type": {
      "description": "The message type, used to identify this as a CAIP-X message",
      "const": "caip-x"
    },
    "data": {

>       "description": "A JSON-RPC message",

Reviewer: Can we be more specific here about the type of message we expect? Or at least give some specific examples, maybe not an exhaustive list. e.g. we could say that this is expected to be an EIP-1193 message OR CAIP-25/27, etc.

Author: done here https://github.com/jiexi/CAIPs/pull/1/commits/6d14374ede59ce8b712ecc36f5f3ef1d6bd49e05

      "type": "object"
    }
  },
  "required": ["type", "data"]
}
```

### `externally_connectable`

A web extension can use the `externally_connectable` manifest field to accept messages from websites and other extensions. This permission can be configured to limit which sites and other extensions can send messages to the web extension. How this permission is configured is out-of-scope for this proposal; this proposal only concerns sites that the web extension allows messages from.

The web extension can:
* handle a connection from the website or other extensions by using `chrome.runtime.onConnectExternal.addListener((port) => {...})`
* send messages by using `port.postMessage()`
* receive messages by using `port.onMessage.addListener()`
  * incoming messages should be validated according to the message format specification above

The webpage embedded provider can:

> * initiate a connection with the web extension by using `port = browser.runtime.connect()`

Reviewer 2: Don't we need to consider the discoverability part?

One of the most important parts of CAIP-282 is the ability to not only detect installed browser extension wallets... but also the ability to list them with an icon, name and url... plus filter them based on CAIP-217 scopes

* send messages by using `port.postMessage()`
* receive messages by using `port.onMessage.addListener()`

### Caveats

> Currently Firefox does not support `externally_connectable` yet, but they are [considering implementing it](https://bugzilla.mozilla.org/show_bug.cgi?id=1319168). Meanwhile, extension wallets on Firefox will need to continue injecting an inpage provider for the dApp.

Reviewer: Nit: We're inconsistent throughout about the terms "website", "dapp", "dApp". What do we want to call sites that connect to us? IIRC we had discussed this for our docs and decided to say 'website" or "app" or something like that. Maybe to make it more accessible or generic (the term "dapp" may not be familiar to everyone, and for those that are familiar, they may not consider many of the sites that use MetaMask to be dapps)

## Rationale
<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->

> While web extensions could realize the benefits of `externally_connectable` without using it with a standardized interface, this would mean that every web extension would have to ship a provider library that worked specifically for their interface and that Dapps would need to import them. It wouldn't be feasible for Dapps to import every single web extension's specific provider implementation, making this approach less viable.

Reviewer: This seems to be a rationale for "having a standard" versus doing the same thing without a standard. That seems atypical. I would expect the rationale section to discuss why we're doing this in the first place, not just why we're standardizing it.

Reviewer: The point made here about inter-operability is totally relevant, but it's just one benefit of many. Maybe we could frame it more like:

* `externally_connectable` allows websites to communicate with wallets without requiring code injection
* This message format allows us to disambiguate these messages from others that extensions and websites might be sending on this same interface
* Using the standard as the message type enables compatibility with any website <=> wallet message formats (such as EIP-1193, CAIP-25, CAIP-27, etc.), including future formats
* Standardizing this allows websites and wallets to be inter-operable in the same way they are today with EIP-1193.

Reviewer: Maybe this is an appropriate place to list the benefits of not injecting code as well, which I see as:
* Reduced extension permissions
* Improved performance
* Improved website compatibility

Maybe also worth explaining how this reduces complexity for dapp developers? I don't know how to express this idea, but it's the part about how we're forcibly injecting code into another persons application, and they need to consider this in how they build and design that application, and when debugging and monitoring and reviewing bug reports and so on. e.g. like some people are out there triaging Sentry reports for errors that supposedly come from their app, but it's our code. It's a mess for website authors, and this simplifies it somewhat.

Web extension inter-operability is the key to enabling generalized provider implementations. Generalized provider implementations allows for convenient adoption by Dapps, leading to more wide spread adoption of the standard as a result.

## Test Cases
<!--Please add test cases here if applicable.-->

## Security Considerations
<!--Please add an explicit list of intra-actor assumptions and known risk factors if applicable. Any normative definition of an interface requires these to be implementable; assumptions and risks should be at both individual interaction/use-case scale and systemically, should the interface specified gain ecosystem-namespace adoption. -->

> `externally_connectable` has seen a decade of usage via extensions on Chrome. It has a strictly better security when compared to postMessage over contentscript.

Reviewer: I don't know that we can say this. I'm not aware of any security differences between the two, or reasons to think one would be "more secure" in some way.

Author: externally_connectable messages can't be read and/or intercepted and replaced like they can with window.postMessage because externally_connectable sends messages directly to the extension. You even have to specify exactly what extension you want to communicate with, i.e. with extensionId.

Or is this more of a privacy thing?

Reviewer: Messages sent via `postMessage` can't be read or intercepted either though (assuming they have a set target of course). These transports are on-par in this respect right?

It sounds like you're talking more about the contentscript permission versus `externally_connectable`, rather than `postMessage` versus `externally_connectable`. i.e. a contentscript is a higher-privileged permission than `externally_connectable`, it gives the extension more superpowers which comes with more risk for the user.

But it's not as though it makes that extension more or less under threat from other parties.

## Privacy Considerations
<!--Please add an explicit list of intra-actor assumptions and known risk factors if applicable. Any normative definition of an interface requires these to be implementable; assumptions and risks should be at both individual interaction/use-case scale and systemically, should the interface specified gain ecosystem-namespace adoption. -->

> Exposing the wallet API over `externally_connectable` opens a migration path towards a bring-your-own provider model in which web extensions can reduce their fingerprint by no longer having to inject their own provider into every webpage.

Reviewer: Probably worth expanding on the privacy considerations of the discovery mechanisms, when that is documented somewhere. That's likely where we'll have more to say.

Though... couldn't `externally_connectable` be used for fingerprinting? I guess it would be no worse than the fingerprinting that would be possible today with injected code or a contentscript that replies to certain messages - someone could reach out to us in any context and use that to help identify a person. But perhaps we should explain this risk and ensure that it's characterized accurately (e.g. as no worse than the status quo, assuming that is correct).

Author: good point. done here https://github.com/jiexi/CAIPs/pull/1/commits/649f8563f85681e9306f95c688ba5d642544358a

## Backwards Compatibility
<!--All CAIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The CAIP must explain how the author proposes to deal with these incompatibilities. CAIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->

> This CAIP does not require discontinuing usage of contentscript. It is RECOMMENDED that wallets start implementing this alternative connection strategy and encouraging it's usage so that the ecosystem can eventually remove contentscript injection entirely. As an optional transitionary step, wallets can migrate their injected provider to start making connections over `externally_connectable` with no user-facing impact (not sure if true since this has caveats).

Reviewer: The framing here seems.... less than ideal. "This CAIP does not require.... It is RECOMMENDED..." We're starting out by dispelling a misconception that the reader might not even have, then promptly moving on to a related-but-different without elaborating.

Maybe let's start by explaining directly:
* How this is incompatible
* How we'd recommend addressing that

Then we can clarify or dispel misconceptions or whatnot later, if we feel that it's still necessary.

Reviewer: We could also mention here how the message format allows extensions that use `externally_connectable` already to keep using it, because these new messages are clearly identifiable.

Probably an extreme edge case, but relevant because such an extension might consider this a breaking change if they thought they'd need to scrap their existing interface.

---
# Every document starts with a front matter in YAML enclosed by triple dashes.
# See https://jekyllrb.com/docs/front-matter/ to learn more about this concept.
caip: <to be assigned>
title: Web extension provider communication
author: Mark Stacey (@Gudahtt), Jiexi Luan (@jiexi)
discussions-to: <URL>
status: Draft
type: Standard
created: 2022-11-28
---

<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

## Simple Summary
<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the CAIP.-->

> The proposal aims to standardize how inpage injected providers communicate with web extensions, making the process more efficient and secure.

Reviewer 3: definitely not injected 😃 
```suggestion
The proposal aims to standardize how web3 enabled pages communicate with web extensions, making the process more efficient and secure.
```

## Abstract
<!--A short (~200 word) description of the technical issue being addressed.-->

> This CAIP establishes how inpage injected providers can communicate with web extensions over `externally_connectable` in an extensible manner.

Reviewer 3: ```suggestion
This CAIP establishes how web3 enabled pages can communicate with web extensions over `externally_connectable` in an extensible manner.
```

> This CAIP establishes how inpage injected providers can communicate with web extensions over `externally_connectable` in an extensible manner.

Reviewer 3: Also summary and abstract seem to be basically the same thing still

Author: addressed here https://github.com/jiexi/CAIPs/pull/1/commits/5f093b1e2a8a598ed27aaf027c48d4a88df036fa

  "description": "A request sent between a provider and a web extension.",
  "type": "object",
  "properties": {
    "type": {
      "description": "The message type, used to identify this as a CAIP-X message",
      "const": "caip-x"
    },
    "data": {
      "description": "A JSON-RPC message",
      "type": "object"
    }
  },
  "required": ["type", "data"]
}
```

TODO: should type be more defined here? Seems weird to specify the envelop but no types it can be used with?
Does `data` need to be a JSON-RPC message specifically?
Should `type` be something besides `caip-x`?

### `externally_connectable`

A web extension can use the `externally_connectable` manifest field to accept messages from websites and other extensions. This permission can be configured to limit which sites and other extensions can send messages to the web extension. How this permission is configured is out-of-scope for this proposal; this proposal only concerns sites that the web extension allows messages from.

The web extension can:
* handle a connection from the website or other extensions by using `chrome.runtime.onConnectExternal.addListener((port) => {...})`

> * send messages by using `port.postMessage()`

Reviewer 3: shouldn't this be [runtime.sendMessage](https://developer.chrome.com/docs/extensions/reference/api/runtime#method-sendMessage)?

Author: connect() is more appropriate for long lived connections, specifically when the extension may send messages that are not responses to a sendMessage from the dapp. 

I don't believe the extension has a runtime.sendMessage equivalent to the dapp

Reviewer 3: ah right ok yes quite right

* receive messages by using `port.onMessage.addListener()`
  * incoming messages should be validated according to the message format specification above

The provider can:
* initiate a connection with the web extension by using `port = browser.runtime.connect()`

> * send messages by using `port.postMessage()`

Reviewer 3: same here?

Reviewer 3: disregard!

> * send messages by using `port.postMessage()`

Reviewer 3: ```suggestion
* send messages by using `port.runtime.sendMessage()`
```

> * receive messages by using `port.onMessage.addListener()`

Reviewer 3: ```suggestion
* receive messages by using `port.runtime.onMessage.addListener()`
```

---
# Every document starts with a front matter in YAML enclosed by triple dashes.
# See https://jekyllrb.com/docs/front-matter/ to learn more about this concept.
caip: <to be assigned>
title: Web extension provider communication
author: Mark Stacey (@Gudahtt), Jiexi Luan (@jiexi)
discussions-to: <URL>
status: Draft
type: Standard
created: 2022-11-28
---

<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

## Simple Summary
<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the CAIP.-->

> The proposal aims to standardize how providers communicate with web extensions, making the process more efficient and secure.

Reviewer 3: I wonder if we should be a bit more specific about what provider means in this context?

Reviewer 3: or perhaps this context can be taken for granted given this is landing in CASA?

Author: we can totally elaborate on what a provider is. You asking this question has me thinking that we should be only focus on providers in so much as they provide a convenience wrapper for wallet communication. That is to say, we should more broadly say "Dapp" in place of "provider" in several places in this document

Reviewer 3: Yeah this is tricky. Provider is a potentially confusing word since it operates as this nebulous thing that "provides" an interface for blockchains and here we're changing our assumption of where this provider lives - previously injected, now BYOP - in the stack without even providing (:wink:) that context.

Author: how about this https://github.com/jiexi/CAIPs/pull/1/commits/4205dc5646d82c65c1901a821f240a09d0d2f83c?

Reviewer 3: adding this comment I left on the commit here:

> 🤔 it wouldn't be injected anymore though?

> The proposal aims to standardize how JSON-RPC providers with web extensions communicate, making the process more efficient and secure.

Author: they don't necessarily have to be JSON-RPC providers, do they? The envelope itself isn't JSON-RPC like either

Reviewer 3: I wouldn't think so

Author: addressed https://github.com/jiexi/CAIPs/pull/1/commits/6641dd411cd6302efbf1b0c81b9941ae819c78ae

## Abstract
<!--A short (~200 word) description of the technical issue being addressed.-->

> This CAIP establishes how providers can communicate with web extensions over `externally_connectable` in an extensible manner.

Reviewer 3: summary and abstract seem be slightly reworded versions of the same thing

## Motivation
<!--The motivation is critical for CAIP. It should clearly explain why the state of the art is inadequate to address the problem that the CAIP solves. CAIP submissions without sufficient motivation may be rejected outright.-->

> Web extensions today will typically inject a JavaScript provider API into websites as a global variable. For example, in the Ethereum ecosystem this provider API is standardized in EIP-1193, and is conventionally injected as `window.ethereum`.

Reviewer 3: ```suggestion
Web extension crypto wallets today will typically inject a JavaScript provider API into websites as a global variable. For example, in the Ethereum ecosystem this provider API is standardized in EIP-1193, and is conventionally injected as `window.ethereum`.
```

Reviewer 3: or just
```suggestion
Web extension wallets today will typically inject a JavaScript provider API into websites as a global variable. For example, in the Ethereum ecosystem this provider API is standardized in EIP-1193, and is conventionally injected as `window.ethereum`.
```

This injected API strategy has some advantages:
* For websites developers, using a global variable is simple and requires no effort on their part to setup.
* It allows websites to support any web extension following this standard with no additional effort.

However, the injected API strategy has many disadvantages:

> * It depends upon the web extension having read and write access to every website, which is a scary permission that web extension authors might otherwise be able to avoid asking for.

Reviewer 3: ```suggestion
* It depends upon the web extension having read and write access to every website the user visits, which is a scary permission that web extension authors might otherwise be able to avoid asking for.
```

* It slows down every website by injecting additional code to be parsed and executed. It even slows down the initial page load in most cases, because many web extensions inject the provider synchronously to maintain compatibility with websites that expect it to be available immediately.

> * Injecting code into a website can break it.

Reviewer 3: ```suggestion
* In some cases injecting code into a website can even break the site.
```

Author: maybe this? https://github.com/jiexi/CAIPs/pull/1/commits/518c27ec7477b1ea85ecde11fd865f9a1d8b9a86

> * It provides no way for web extension authors to safely make breaking changes to their provider API, resulting in even more code being injected to support legacy APIs.

Reviewer 3: maybe we should explain this one a bit more?

Author: tbh, I'm not sure I fully understand it myself haha

Author: @Gudahtt

Reviewer 3: I guess I get what its driving at but I'm not clear on how `externally_connectable` resolves the issue of breaking changes other than no longer needing "more code" to be injected

Reviewer 3: Oh I see its answered directly below: https://github.com/jiexi/CAIPs/pull/1/files#diff-c975f3fd9fb25a0964359ba4a49e22edd4fac12cc60d0645ee191f6a02b14069R42

Author: reworded here https://github.com/jiexi/CAIPs/pull/1/commits/a2c083a6246d221786366f5cd7d839958bfe04f7

An alternative strategy would be for a website to embed its own provider. A provider could be offered as a library, to be embedded by the website author. This strategy can address all disadvantages of the injected provider approach:
* If this strategy became widespread enough, it would allow some web extensions to stop asking for write access to all pages.

>   * NOTE: Should mention discovery somewhere since that will require injection still

Author: TODO

Reviewer 3: I've got a parallel CAIP WIP. Will link here shortly

Author: https://github.com/adonesky1/CAIPs/pull/1

* The website author can control when the provider is initialized, and fine-tune performance.
* No code needs to be injected, so websites would no longer be broken by injected code.
* The provider library can be published with breaking changes, allowing changes to the API without needing to embed legacy code in each website.

A provider library can be similarly easy to use for website authors as well, requiring nothing more than a single script tag to import a library and get an equivalent experience to using an the injected provider.

> Web extension inter-operability is a challenge for embeded providers though. That is what this proposal means to address. Today there is no way to write a provider such that it is compatible with any web extension. Web extensions differ today in how they communicate with wallets, from the messaging system used to the messaging format. These details often aren't publicly documented or treated as a public-facing API, so they can change without notice, making it risky even to embed support for popular conventions used today.

Reviewer 3: ```suggestion
Web extension inter-operability is a challenge for embedded providers though. That is what this proposal means to address. Today there is no way to write a provider such that it is compatible with any web extension. Web extensions differ today in how they communicate with wallets, from the messaging system used to the messaging format. These details often aren't publicly documented or treated as a public-facing API, so they can change without notice, making it risky even to embed support for popular conventions used today.
```

> A standard method for providers to communicate with web extebsions would allow website authors to embed their own providers without losing web extension inter-operability.

Reviewer 3: ```suggestion
A standard method for providers to communicate with web extensions would allow website authors to embed their own providers without losing web extension inter-operability.
```

"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
uppercase in this document are to be interpreted as described in [RFC
2119](https://www.ietf.org/rfc/rfc2119.txt)

### Summary

Web extensions should expose a standard interface over [`externally_connectable`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable) to enable the inter-operability of embedded providers.

### Message format

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "/caip-x/schemas/provider-request.schema.json",
  "title": "Provider / Web Extension message",
  "description": "A request sent between a provider and a web extension.",
  "type": "object",
  "properties": {
    "type": {
      "description": "The message type, used to identify this as a CAIP-X message",
      "const": "caip-x"
    },
    "data": {
      "description": "A JSON-RPC message",
      "type": "object"
    }
  },
  "required": ["type", "data"]
}
```

TODO: should type be more defined here? Seems weird to specify the envelop but no types it can be used with?
Does `data` need to be a JSON-RPC message specifically?

> Should `type` be something besides `caip-x`?

Author: Something missing from this CAIP is a specification for what the type for an EVM provider should look like, and the shape of its data

Author: not sure if data needs to be an object

Author: I think we need yet another CAIP that specifies what the shape of evm provider data that goes through this envelope

Reviewer 3: > I think we need yet another CAIP that specifies what the shape of evm provider data that goes through this envelope

I'm not entirely sure this is necessary, but am open to it

Author: actually hmm... so currently we envelope our port stream connection messages, but that is a MetaMask specific implementation detail. I'm not sure that it is something we would want to include in a standard. Perhaps this was Mark's intention and I am only just now understanding/realizing that. This would mean any MetaMask specific messages/streams are still necessary to be flattened and contained within JSON-RPC formatted messages instead. I imagine this won't be an issue, but haven't fully investigated yet

Reviewer 3: Where are we on this thread now?

Author: need to update this section to specify what an ethereum provider message should look like for type `caip-x`. `caip-x` should be the caip number that this proposal gets assigned. Future CAIPs can extend this

Reviewer 3: I'm still not clear on why we need
> to specify what an ethereum provider message should look like for type caip-x

?
Our provider will no longer be specific to ethereum? And shouldn't mirror 1193 provider functionality?

Author: I believe the idea was that the `type` field would specify the `data` format

So for Ethereum it maybe `caip-100` with JSON-RPC in the data field. For Cardono it maybe `caip-101` with GraphQL query in the data field

### `externally_connectable`

A web extension can use the `externally_connectable` manifest field to accept messages from websites and other extensions. This permission can be configured to limit which sites and other extensions can send messages to the web extension. How this permission is configured is out-of-scope for this proposal; this proposal only concerns sites that the web extension allows messages from.

The web extension can:
* handle a connection from the website or other extensions by using `chrome.runtime.onConnectExternal.addListener((port) => {...})`
* send messages by using `port.postMessage()`
* receive messages by using `port.onMessage.addListener()`
  * incoming messages should be validated according to the message format specification above

The provider can:
* initiate a connection with the web extension by using `port = browser.runtime.connect()`
* send messages by using `port.postMessage()`
* receive messages by using `port.onMessage.addListener()`

Caveats:
* No firefox support yet, but they are considering implementing it
* Extension Ids may be different across different family of browsers
  * Discoverability should address this

> 

Author: not sure where to put these quite yet.

Also need to flesh out Extension IDs being different? Or do we think we can push that into the discoverability CAIP?

Author: Still unsure where this should live

Reviewer 3: > Also need to flesh out Extension IDs being different? Or do we think we can push that into the discoverability CAIP?

yeah atleast on this point I can bring this into [the WIP discovery CAIP](https://github.com/adonesky1/CAIPs/pull/1)

Reviewer 3: > No firefox support yet, but they are considering implementing it

This seems like the right place to mention this. Maybe we can find the right link to point to their indications of eventually supporting it

Author: both points addressed here https://github.com/jiexi/CAIPs/pull/1/commits/e7a8390561715af4f3cb8e6705a9fc94bf8d8ce4

leaving the different extension IDs to the discovery CAIP

---
# Every document starts with a front matter in YAML enclosed by triple dashes.
# See https://jekyllrb.com/docs/front-matter/ to learn more about this concept.
caip: <to be assigned>
title: Web extension provider communication
author: Mark Stacey (@Gudahtt), Jiexi Luan (@jiexi)
discussions-to: <URL>
status: Draft
type: Standard
created: 2022-11-28
---

<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

## Simple Summary
<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the CAIP.-->

> This CAIP discusses the motivation, specification, and rationale for a proposal aimed at improving how web extension wallets interact with websites, particularly focusing on the Ethereum ecosystem. It outlines the current method of injecting JavaScript provider APIs into websites, its advantages, and its numerous disadvantages, such as security concerns, performance issues, and the risk of breaking websites. An alternative strategy is proposed that specifies a standard communication specification over a new transport layer which enables websites to be able to embed their own provider as a library, addressing the disadvantages of injecting providers into websites and improving web extension interoperability as a whole.

Reviewer 3: not sure why this should be focused on ethereum? Other than noting that the previous model we used was explicitly Ethereum oriented? 
```suggestion
This CAIP discusses the motivation, specification, and rationale for a proposal aimed at improving how web extension wallets interact with websites. It outlines the current method of injecting JavaScript provider APIs into websites, its advantages, and its numerous disadvantages, such as security concerns, performance issues, and the risk of breaking websites. An alternative strategy is proposed that specifies a standard communication specification over a new transport layer which enables websites to be able to embed their own provider as a library, addressing the disadvantages of injecting providers into websites and improving web extension interoperability as a whole.
```

      "const": "caip-x"
    },
    "data": {
      "description": "A JSON-RPC message",
      "type": "object"
    }
  },
  "required": ["type", "data"]
}
```

TODO: should type be more defined here? Seems weird to specify the envelop but no types it can be used with?
Does `data` need to be a JSON-RPC message specifically?
Should `type` be something besides `caip-x`?

### `externally_connectable`

A web extension can use the `externally_connectable` manifest field to accept messages from websites and other extensions. This permission can be configured to limit which sites and other extensions can send messages to the web extension. How this permission is configured is out-of-scope for this proposal; this proposal only concerns sites that the web extension allows messages from.

The web extension can:
* handle a connection from the website or other extensions by using `chrome.runtime.onConnectExternal.addListener((port) => {...})`
* send messages by using `port.postMessage()`
* receive messages by using `port.onMessage.addListener()`
  * incoming messages should be validated according to the message format specification above

> The provider can:

Reviewer 3: maybe a bit clearer:
```suggestion
The webpage embedded provider can:
```
